augur/filter/io.py

```diff
@@ -96,46 +99,85 @@ def constant_factory(value):
         raise AugurError(f"missing or malformed priority scores file {fname}")
 
 
-def write_metadata_based_outputs(input_metadata_path: str, delimiters: Sequence[str],
-                                 id_columns: Sequence[str], output_metadata_path: str,
-                                 output_strains_path: str, ids_to_write: Set[str]):
+def get_cat(file):
+    if file.endswith(".gz"):
+        return which("gzcat")
+    if file.endswith(".xz"):
+        return which("xzcat")
+    if file.endswith(".zst"):
+        return which("zstdcat")
+    else:
+        return which("cat")
+
+
+def write_metadata(input_metadata_path: str, delimiters: Sequence[str],
+                   id_columns: Sequence[str], output_metadata_path: str,
+                   ids_to_write: Set[str]):
     """
-    Write output metadata and/or strains file given input metadata information
+    Write output metadata file given input metadata information
     and a set of IDs to write.
     """
     input_metadata = Metadata(input_metadata_path, delimiters, id_columns)
 
-    # Handle all outputs with one pass of metadata. This requires using
-    # conditionals both outside of and inside the loop through metadata rows.
-
-    # Make these conditionally set variables available at this scope.
-    output_metadata_handle = None
-    output_metadata = None
-    output_strains = None
-
-    # Set up output streams.
-    if output_metadata_path:
-        output_metadata_handle = xopen(output_metadata_path, "w")
-        output_metadata = csv.DictWriter(output_metadata_handle, fieldnames=input_metadata.columns,
-                                         delimiter="\t", lineterminator=os.linesep)
-        output_metadata.writeheader()
-    if output_strains_path:
-        output_strains = open(output_strains_path, "w")
-
-    # Write outputs based on rows in the original metadata.
-    for row in input_metadata.rows():
-        row_id = row[input_metadata.id_column]
-        if row_id in ids_to_write:
-            if output_metadata:
-                output_metadata.writerow(row)
-            if output_strains:
-                output_strains.write(row_id + '\n')
-
-    # Close file handles.
-    if output_metadata_handle:
-        output_metadata_handle.close()
-    if output_strains:
-        output_strains.close()
+    output_is_tsv = output_metadata_path.endswith(".tsv")
+    tsv_join = which("tsv-join")
+    cat = get_cat(input_metadata_path)
+
+    # TODO: support compressed outputs when xopen supports them.
+    # Current error when replacing open() with xopen() and using .xz for output:
+    #     'PipedXzWriter' object has no attribute 'fileno'
+    # FIXME: create an issue in the xopen repo
+    if output_is_tsv and tsv_join and cat:
```

Comment:
My thinking here:

1. we'd only benefit from this in cases where the metadata file is large
2. when the metadata file is large, the output type is likely to be compressed

Given those assumptions, there's no point in implementing this without support for compressed outputs. So the next step would be to figure out how to get that working. I tried a bit today and didn't get anywhere, so I think the next step is to open an issue in the xopen repo to support compressed outputs.

Reply:
I think xopen v2 might support fileno now?

Reply:
Just tried using latest xopen (2.0.2) and it's still unsupported with a similar error:

```
io.UnsupportedOperation: fileno
```

Reply:
Issue created: https://github.com/pycompression/xopen/issues/162

Reply:
@corneliusroemer As I read about it, xopen supports _inputs_ that are file objects now, but doesn't always itself return a file object with a fileno.

@victorlin As part of my WIP for `augur merge` needing similar functionality, I actually looked into xopen's support for fileno two weeks ago. Last week, when I saw this, I wrote a comment here, but it never posted because I didn't realize I still had a pending review open (that I accidentally never clicked "submit review" on! :facepalm:). In any case, see e56a91c9b5c3ad643584b4d6bc56456283321194 and https://github.com/nextstrain/augur/commit/74001d211b91248ee8e8dbd16e318ddee2db5ac5#diff-d032c8de9ed6a70822d4a862a85f9e6aaff140bfd8e377d2d09a7a77fca73a07R152. We can talk it thru too.

Note that this isn't really an xopen issue, or at least, I wouldn't consider it one and I suspect the author wouldn't either. xopen's scope is providing _Python file objects_, not arbitrary file descriptors you can pass/hook up to another process (as Popen requires). While it's possible (with a couple big caveats) to dig into the internals of the file objects provided by xopen and obtain a file descriptor most (but not all) of the time, it's a path I decided wasn't worth the caveats in the end for `augur merge`.

Reply:
Re: xopen fileno support, perhaps the most relevant comment in my `augur read-file` implementation is:

> Note that even in the common case of `xopen()` reading from a pipe _[which has a fileno]_—the stdout of an external decompression process—we can't use that pipe directly because it always [buffers the first block](https://github.com/pycompression/xopen/blob/67651844/src/xopen/__init__.py#L374) of the file into Python.

Reply:
Updated to use `copyfileobj` (inspired by https://github.com/nextstrain/augur/pull/1562):

https://github.com/nextstrain/augur/blob/91dafbf95aa2ee200c794a986cfdd570c6470d7e/augur/filter/io.py#L148-L151